### ISSUE TITLE
Flow null detection resulting in false positives

### ISSUE THREAD

**Issue by mqualizz** [facebook/flow#6898]
I've tested this with flow 0.81.  

https://flow.org/try/#0GYVwdgxgLglg9mABAcwKZQPICMBWroAUAlAFwD8YIAtlqgE6IDeAUIm4hAgM5SICGiALyIAsnygALAHR0+YACZwqxANytE6tnXQg6SAgIA8iKQFYiiMv0QlElADb21AX2bNQkWAkRRUPYkyaiPbo1sJomLj4UKpBIbxYQijo2HiERGpBMMCIBkLCDvaIAGTFiImCBSCOFizs9YjaULpgavWubFk5eQCEVY4lZYl9dtX2tUH12bkClcJYEw1L5UmFbcsd7JtdM-mjNYHLYeXrSxX7TkHbU90V-eOHy+cCAFSIAEyniNfsnGBccBCUnscGQBAABgASRh8ZzgjJBP4AoEgsFQxhYOEI1zMXz+DJAA

Flow is not recognizing that some values should not be null.  Errors on lines 27 and 30.

function getObject():?number {
    const a = Math.random();
  
    return (a < .5) ? a : null;
}

function test() {
    let a = getObject();
    let b = getObject();

    if (a == null && b == null) {
        return;
    }
  
    if (a != null && b != null) {
        if (a === b) {
            b = null;
        }
    }
  
    if (a == null) {
        a = b;
        b = null;
    }
  
    if (b == null) {
/* 27 */        b = a * 2;
    }
  
/* 30 */    console.log(`${a}`);
    console.log(`${b}`);
}

test();

**Comment by wchargin** [comment]
Flow tracks a type for each variable. It does not track conditions over
_multiple_ variables, like “either `a` or `b` might be null, but we know
that they are not _both_ null”.

The reason that `a` is guaranteed to be non-null after line 24 depends
on this complex condition. At that point, Flow knows that `a` is null
and `b` is a `?number`, and you assign `a = b`. From this information,
it is not guaranteed that the new value of `a` is non-null. The
guarantee only comes from line 22, where you establish the condition
that `a` and `b` are not simultaneously null.

Inserting `if (b == null) throw new Error("unreachable");` before
line 23 resolves the type errors.

**Comment by mqualizz** [comment]
That makes sense, thank you.

**Comment by mqualizz** [comment]
I have another "false positive" scenario:

Given:

```
let a:?number = 1;
  
console.log(`${a}`);
```

Flow complains with:

```
3: console.log(`${a}`);
                  ^ Cannot coerce `a` to string because null or undefined [1] should not be coerced.
References:
1: let a:?number = 1;
         ^ [1]
```

Shouldn't Flow be able to determine that, at line 3, a is definitely not null?

**Comment by wchargin** [comment]
> ```js
> let a:?number = 1;
>
> console.log(`${a}`);
> ```
>
> Shouldn't Flow be able to determine that, at line 3, a is definitely
> not null?

Nope. You’ve explicitly told Flow that the type of `a` is `?number`,
which permits `a` to be `null`. If you omit the type annotation on `a`,
then Flow will infer the stronger type `number` and your example will
typecheck.

**Comment by wchargin** [comment]
You’re quite welcome.

What you call “type tightening” is generally called _refinement_. When
you have an explicit condition, like an equality or `instanceof` check,
Flow can refine the type of the value being checked within the scope of
the `if`-statement.

So, when you write `if (a === 1) { /* … */ }`, inside the braces the
type of `a` is refined to `1`, which is known to be non-null.

You’re correct that Flow doesn’t override your declared types by
refining them against their actual values. This is critical for
encapsulation. For instance, suppose that I have:

```js
// zoo.js
export const fido: Animal = new Dog("fido");
```

The module `zoo` tells its clients that they can access `fido`, and that
`fido` is an `Animal`. Clients are told that `fido` is specifically a
`Dog`—for all they know, `fido` could be a zebra. Therefore, clients can
only perform operations common to all animals.

This is useful because it means that the `zoo` module can later change
its implementation to

```js
// zoo.js
export const fido: Animal = new Zebra("fido");
```

without breaking backward compatibility. Information hiding—limiting
implementation details to the local places where they’re strictly
needed—is extremely useful for building large, pluggable systems.
